internal/match/json.go

```diff
@@ -14,21 +15,43 @@ type JSON func(body []byte) error
 
 // JSONKeyEqual returns a matcher which will check that `wantKey` is present and its value matches `wantValue`.
 // `wantKey` can be nested, see https://godoc.org/github.com/tidwall/gjson#Get for details.
-// `wantValue` is matched via reflect.DeepEqual and the JSON takes the forms according to https://godoc.org/github.com/tidwall/gjson#Result.Value
+// `wantValue` is matched via LenientDeepEqual and the JSON takes the forms according to https://godoc.org/github.com/tidwall/gjson#Result.Value
 func JSONKeyEqual(wantKey string, wantValue interface{}) JSON {
 	return func(body []byte) error {
 		res := gjson.GetBytes(body, wantKey)
 		if !res.Exists() {
 			return fmt.Errorf("key '%s' missing", wantKey)
 		}
 		gotValue := res.Value()
-		if !reflect.DeepEqual(gotValue, wantValue) {
-			return fmt.Errorf("key '%s' got '%v' want '%v'", wantKey, gotValue, wantValue)
+		if !LenientDeepEqual(gotValue, wantValue) {
+			return fmt.Errorf(
+				"key '%s' got '%v' (type %T) want '%v' (type %T)",
+				wantKey, gotValue, gotValue, wantValue, wantValue,
+			)
+		} else {
+			return nil
 		}
-		return nil
 	}
 }
 
+// LenientDeepEqual is a version of reflect.DeepEqual that'll be more lenient with the second value,
+// converting it to json-unmarshal canonical form is necessary to complete a valid equality operation.
```

Comment:
This is supposed to be "lenient" towards internal golang values, because to deep equal, a float64 is not the same as float32, or uint, or int

The from-the-hip bespoke value creation passed to json value matching often doesn't match the exact JSON-to-value value that deepequal is checking against, you can see that with https://github.com/matrix-org/complement/issues/278, where every value from JSON is a float64, even if matrix doesn't allow or have floats in endpoints.

Ditto that a `[]interface{}{""}` is not the same as a `[]string{""}`, to deepequal, while it is effectively "the same".

---

Because of this, the easiest way to align the types (recursively) of the second value to the first, is to pull it through a marshal-unmarshal cycle, making it easier to compose values and reducing bugs due to subtle type mismatches.